lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+
+// ignore_for_file: prefer_generic_function_type_aliases
+library over_react.component_declaration.function_component;
+
+import 'package:js/js_util.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+import 'package:react/react.dart' as react;
+import 'package:react/react_client.dart';
+import 'package:react/react_client/js_backed_map.dart';
+
+export 'component_type_checking.dart'
+    show isComponentOfType, isValidElementOfType;
+
+/// A React component declared using a function that takes in [props] and returns rendered output.
+///
+/// See <https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components>.
+typedef dynamic UiFunctionComponent<T extends UiProps>(T props);
+
+/// Returns a `UiFactory<TProps>` for [functionComponent].
+///
+/// __Example__:
+/// ```dart
+/// UiFactory<FooProps> Foo = uiFunction((props) {
+///     // Set default props using null-aware operators.
+///     final isDisabled = props.isDisabled ?? false;
+///     final items = props.items ?? [];
+///
+///     // Return the rendered component contents here.
+///     return Dom.div()(items);
+///   },
+///   // The generated props config will match the factory name.
+///   $FooPropsConfig,
+/// );
+///
+/// mixin FooProps on UiProps {
+///   bool isDisabled;
+///   Iterable<String> items;
+/// }
+/// ```
+///
+/// __OR__ Optionally pass in an existing [PropsFactory] in place of a props [config].
+///
+/// ```dart
+/// UiFactory<FooProps> Bar = uiFunction((props) {
+///     return Dom.div()(props.items);
+///   },
+///   null,
+///   propsFactory: PropsFactory.fromUiFactory(Foo),
+///   displayName: 'Bar',
+/// );
+/// ```
+///
+/// __OR__ Set [config] to `null` when using `UiProps`.
+///
+/// ```dart
+/// UiFactory<UiProps> Baz = uiFunction((props) {
+///     return Dom.div()('prop id: ${props.id}');
+///   },
+///   null,
+///   displayName: 'Baz',
+/// );
+/// ```
+///
+/// Learn more: <https://reactjs.org/docs/components-and-props.html#function-and-class-components>.
+// FIXME right now only top level factory declarations will generate props configs.
+UiFactory<TProps> uiFunction<TProps extends UiProps>(
+  UiFunctionComponent<TProps> functionComponent,
+  FunctionComponentConfig<TProps> config, {
+  PropsFactory<TProps> propsFactory,
+  String displayName,
+}) {
+  if (config != null) {
+    if (propsFactory != null) {
+      throw ArgumentError('propsFactory cannot be used along with config');
+    }
+    propsFactory = config.propsFactory;
+    displayName ??= config.componentName;
+  }
+
+  // Get the display name from the inner function if possible so it doesn't become `_uiFunctionWrapper`
+  displayName ??= _getFunctionName(functionComponent);
+
+  dynamic _uiFunctionWrapper(Map props) {
+    return functionComponent(propsFactory.jsMap(props as JsBackedMap));
+  }
+
+  ReactDartFunctionComponentFactoryProxy factory =
+      react.registerFunctionComponent(
```

Comment:
#nit prefer omitting LHS typing for local variables
```suggestion
  final factory = react.registerFunctionComponent(
```